### PR TITLE
Issue 7131 Implement Newtonian search for lookup

### PR DESCRIPTION
**Pull request by amit-kumar59**
**Change log description**  
In order to support efficient lookups in the index segment we should implement a version of Newtonian search to locate the desired index entry.

**Purpose of the change**  
we want to search over an index segment, which will contain times stamp, offset pairs for, a given segment through the Newtonian Search

**What the code does**  
(Detailed description of the code changes)

**How to verify it**  
(Optional: steps to verify that the changes are effective)

### CODE REVIEW

File: common/src/main/java/io/pravega/common/util/NewtonianSearch.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.common.util;
+import java.util.function.Function;
+
+/**
+ * TO-DO.
```
Comment (tkaitchuck) [inline common/src/main/java/io/pravega/common/util/NewtonianSearch.java:20]: Add docs

File: common/src/main/java/io/pravega/common/util/NewtonianSearch.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ */
+public class NewtonianSearch {
+    @SuppressWarnings("unchecked")
```
Comment (tkaitchuck) [inline common/src/main/java/io/pravega/common/util/NewtonianSearch.java:23]: Add docs

File: common/src/main/java/io/pravega/common/util/NewtonianSearch.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.common.util;
+import java.util.function.Function;
+
+/**
+ * TO-DO.
+ */
+public class NewtonianSearch {
```
Comment (tkaitchuck) [inline common/src/main/java/io/pravega/common/util/NewtonianSearch.java:22]: Please add this to an appropriate class in common.

File: common/src/main/java/io/pravega/common/util/NewtonianSearch.java (hunk 2 above)
Comment (tkaitchuck) [inline common/src/main/java/io/pravega/common/util/NewtonianSearch.java:23]: These warnings that are being suppressed are real errors in the code. Please don't ignore lints like this.

File: common/src/main/java/io/pravega/common/util/NewtonianSearch.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        return value;
+    }
+
+    private static <T, R extends Comparable<R>> int interpolate(int low, int high, R lowValue, R highValue,
+                                                                R targetValue) {
+        return low
+                + (int) (((double) (targetValue.compareTo(lowValue)) / (highValue.compareTo(lowValue))) * (high - low));
+    }
+
+}
```
Comment (tkaitchuck) [inline common/src/main/java/io/pravega/common/util/NewtonianSearch.java:55]: Please add tests which verify:
* Correctness
* Number of lookups compared to binary search
* What happens with invalid input

File: common/src/main/java/io/pravega/common/util/NewtonianSearch.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            int cmp = midValue.compareTo(targetValue);
+            if (cmp < 0) {
+                low = mid + 1;
+            } else if (cmp > 0) {
+                high = mid - 1;
+            } else {
+                return mid;
+            }
+        }
+        return -(low + 1);
```
Comment (tkaitchuck) [inline common/src/main/java/io/pravega/common/util/NewtonianSearch.java:42]: The semantic of returning a negative value is not clearly expressed in the method signature. As this feature is not needed we should probably omit it.

Comment (tkaitchuck) [inline common/src/main/java/io/pravega/common/util/NewtonianSearch.java:42]: By returning the index the caller must re-query the valueExtractor or else cache the results on their own. Hence it is better to return both the index and value.

**Review by tkaitchuck — CHANGES_REQUESTED** [review]

**Comment by amit-kumar59** [comment]
Closing this PR as respective changes are incorporated in PR https://github.com/pravega/pravega/pull/7165

**Pull request closed without merging**